packages/runtime/lib/runtime.js

```diff
@@ -1747,35 +1748,39 @@ class Runtime extends EventEmitter {
     for (const raw of data.split('\n')) {
       // First of all, try to parse the message as JSON
       let message
+      let json
       // The message is a JSON object if it has at least 2 bytes
       if (raw.length >= 2) {
         try {
           message = JSON.parse(raw)
-        } catch (e) {
+          json = true
+        } catch {
           // No-op, we assume the message is raw
         }
       }
 
-      // Not a Pino JSON, accumulate the message and continue
-      if (typeof message?.level !== 'number' || (typeof message?.time !== 'number' && typeof message?.msg !== 'string')) {
-        plainMessages += (plainMessages.length ? '\n' : '') + raw
+      const pinoLog = typeof message?.level === 'number' && typeof message?.time === 'number' && typeof message?.msg === 'string'
+
+      // Directly write to the Pino destination
+      if (pinoLog) {
+        this.#loggerDestination.lastLevel = message.level
+        this.#loggerDestination.lastTime = message.time
+        this.#loggerDestination.lastMsg = message.msg
+        this.#loggerDestination.lastObj = message
+        this.#loggerDestination.lastLogger = logger
+        this.#loggerDestination.write(raw + '\n')
         continue
       }
 
-      // Before continuing, write any previous plain messages
-      if (plainMessages.length > 0) {
-        logger[level]({ caller }, plainMessages.replace(/\n$/, ''))
-        plainMessages = ''
+      if (json) {
+        logger[level]({ caller, [label]: message })
```

Comment:
@mcollina since we have it there the `raw` string, is there a way in pino to pass the raw string and avoid to stringify it again?